Add new rules for aliases, maps, lists and 'with do' (#2)
jungsoft/elixir-style-guide: README.md

<img src="https://jungsoft.io/static/media/jungsoft_logo.c44eaf52.png" width="300px"/>

# Elixir Style Guide

Welcome to Jungsoft's Elixir Style Guide. Sit down, have a cup of coffee and read this calmly. ☕

## Indroduction

The goal of this guide is to help our team to understand and
follow code style best practices, maintaining a pattern.

As this guide is an extension of the [Christopher Adams Elixir Style Guide](https://github.com/christopheradams/elixir_style_guide), we **highly recommend** reading it before you continue.

# Table of Contents

* [Alias Organization](#alias-organization)
* ['With do' Organization](#with-do-organization)
* [Maps](#maps)
* [Lists](#lists)

## Aliases

> ### Comma

Reviewer: We're kind of repeating the same rule everywhere, maybe it should be a separate rule, called "Dangling commas" where we say that all enumerables that span multiple lines should have dangling commas. That means:

- aliases
- keyword lists
- lists
- maps

Aliases with multiple lines should contains a comma in the last module.

Example:

```elixir
# ❌ Bad
alias Jungsoft.{
  Module1,
  Module2
}

# ✅ Good
alias Jungsoft.{
  Module1,
  Module2,
}
```

### Ordering

The aliases should be ordered alphabetically.

Example:

```elixir
# ❌ Bad
alias Jungsoft.{
  CModule,
  BModule,
  AModule,
}

# ✅ Good
alias Jungsoft.{
  AModule,
  BModule,
  CModule,
}
```

### SubModules

> The aliases should contains only one 'sub-module'.

Reviewer: ```suggestion
The aliases should contain only one 'sub-module'.
```

Example:

```elixir
# ❌ Bad
alias Jungsoft.{
  Module1.SubModule1,
  Module1.SubModule2,
  Module2.SubModule1.SubSubModule1,
  Module2.SubModule1.SubSubModule2,
}

# ✅ Good
alias Jungsoft.Module1.{
  SubModule1,
  SubModule2,
}
alias Jungsoft.Module2.SubModule1.{
  SubSubModule1,
  SubSubModule2,
}
```

[Back to top ⬆️](#table-of-contents)

## 'With do' Organization

### Space

When `with do` contains multiple lines, it's necessary to add two blank spaces before the first parameter. This is only valid when editor tab size is **2**. If you followed the Jungsoft guidelines this should be your case. 😊

Example:

```elixir
# ❌ Bad
with true <- Jungsoft.function_1(),
      false <- Jungsoft.function_2()
do
  ## DO SOMETHING
end

# ❌ Bad
with true <- Jungsoft.function_1(),
     false <- Jungsoft.function_2()
do
  ## DO SOMETHING
end

# ❌ Bad
with  true <- Jungsoft.function_1() do
  ## DO SOMETHING
end

# ✅ Good
with  true <- Jungsoft.function_1(),
      false <- Jungsoft.function_2()
do
  ## DO SOMETHING
end

# ✅ Good
with true <- Jungsoft.function_1() do
  ## DO SOMETHING
end
```

> ### Blank lines

Reviewer: Maybe this rule should be more generic? It's not specific to `with`, but we should avoid unnecessary blank lines everywhere. For example:

```elixir
# ❌ Bad
defmodule Module do
  @moduledoc false

  def func do

    1 + 1
  end

end

# ✅ Good
defmodule Module do
  @moduledoc false

  def func do
    1 + 1
  end
end
```

The `with do` parameters should not be separated by blank spaces. Except when `case` contains multiple parameters. In this case should use the same logic as `case` and `cond` normal style (see [here](https://github.com/christopheradams/elixir_style_guide#multiline-case-clauses)).

Example:

```elixir
# ❌ Bad
with  %StarWars{} <- Jungsoft.function_1(),
      true <- Jungsoft.function_2()
do

  ## DO SOMETHING

else

  %StarTrek{} ->
    {:error, "Wrong choice"}

  false ->
    ## DO SOMETHING

end

# ✅ Good
with  %StarWars{} <- Jungsoft.function_1(),
      true <- Jungsoft.function_2()
do
  ## DO SOMETHING
else
  %StarTrek{} ->
    {:error, "Wrong choice"}

  false ->
    ## DO SOMETHING
end
```

[Back to top ⬆️](#table-of-contents)

## Maps

### Comma dangle

When map contains multiple lines, the last line should contain a comma.

Example:

```elixir
# ❌ Bad
map = %{
  first: "first",
  second: "second".
  third: "third"
}

# ✅ Good
map = %{
  first: "first",
  second: "second".
  third: "third",
}

```

[Back to top ⬆️](#table-of-contents)

## Lists

### Maximum parameters

> When a list contains more than 3 values, it should be breaking into multiple lines.

Reviewer: ```suggestion
When a list contains more than 3 values, it should be broken into multiple lines.
```

> Example:

Reviewer: this should apply to all enumerables, right?

Example:

```elixir
# ❌ Bad
map = %{a: 1, b: 2, c: 3, d: 4}

# ✅ Good
map= %{
  a: 1,
  b: 2,
  c: 3,
  d: 4,
}

```elixir
# ❌ Bad
list = ["value1", "value2", "value3", "value4"]

# ✅ Good
list = [
  "value1",
  "value2",
  "value3",

>   "value4"

Reviewer: This should also respect the dangling comma rule

```suggestion
  "value4",
```

<img src="https://jungsoft.io/static/media/jungsoft_logo.c44eaf52.png" width="300px"/>

# Elixir Style Guide

Welcome to Jungsoft's Elixir Style Guide. Sit down, have a cup of coffee and read this calmly. ☕

## Indroduction

The goal of this guide is to help our team to understand and
follow code style best practices, maintaining a pattern.

As this guide is an extension of the [Christopher Adams Elixir Style Guide](https://github.com/christopheradams/elixir_style_guide), we **highly recommend** reading it before you continue.

# Table of Contents

* __[Formatting](#formatting)__
  * [Breaking Lines](#breaking-lines)
  * [Blank Lines](#blank-lines)
  * [Dangling Commas](#dangling-commas)
* __[Schemas](#schemas)__
  * [Ecto Schema](##ecto-schema)
  * [GraphQL Schema](##graphql-schema)
* __[Regular Expressions](#regular-expressions)__
  * [Aliases](#aliases)
  * ['With do' Organization](#with-do-organization)

# Formatting
## Breaking Lines

When an enumerable contains more than 3 arguments, it should be broken into multiple lines. See an example for Lists below.

```elixir
# ❌ Bad

> list = ["value1", "value2", "value3", "value4"]

Reviewer: I think we should also add a map example to make this clearer

# ✅ Good
list = [
  "value1",
  "value2",
  "value3",
  "value4",
]

# ✅ Good
list = ["value1", "value2", "value3"]
```

**This is also valid for Maps and Keyword Lists.**

[Back to top ⬆️](#table-of-contents)

## Blank Lines

> Blank lines are important in some cases, but in some others can make the code more unreadable.

Reviewer: ```suggestion
Blank lines are important in some cases, but in some others can make the code less readable.
```

  @moduledoc false

  def foo do
    1 + 1
  end
end
```

#### With

The `with do` parameters should not be separated by blank spaces. Except when `else` contains multiple options. In this case should use the same logic as `case` and `cond` normal style (see [here](https://github.com/christopheradams/elixir_style_guide#multiline-case-clauses)).

```elixir
# ❌ Bad
with  %StarWars{} <- Jungsoft.function_1(),
      true <- Jungsoft.function_2()
do

  ## DO SOMETHING

else

  %StarTrek{} ->
    {:error, "Wrong choice"}

  false ->
    ## DO SOMETHING

end

# ✅ Good
with  %StarWars{} <- Jungsoft.function_1(),
      true <- Jungsoft.function_2()
do
  ## DO SOMETHING
else
  %StarTrek{} ->
    {:error, "Wrong choice"}

  false ->
    ## DO SOMETHING
end
```

[Back to top ⬆️](#table-of-contents)

## Dangling Commas

Objects with the possibility to span in multiple lines should contain a comma dangle in the last parameter. See an alias example below.

```elixir
# ❌ Bad
alias Jungsoft.{
  Module1,
  Module2
}

# ✅ Good
alias Jungsoft.{
  Module1,
  Module2,
}
```

**This is also valid for Maps, Lists and Keyword Lists.**

[Back to top ⬆️](#table-of-contents)

> # Schemas

Reviewer: Since these are specific to libraries (ecto and absinthe), I think this section should be the last one, giving more focus to the general elixir language guidelines.

  field :age, :integer

  has_one :one, Bar

  has_one :two, Module
  timestamps()
end

# ✅ Good
schema "foo" do
  field :name, :string
  field :age, :integer

  has_one :one, Bar
  has_one :two, Module

  timestamps()
end
```

The other patterns are different based on the type of schema (Ecto or GraphQL). These differences are listed below.

## Ecto Schema

The Ecto schema parameters need to be organized in a order of importance:

| Importance | Field |
|---|---|
| 1º | field |
| 2º | belongs_to |
| 3º | has_one |
| 4º | has_many |
| 5º | timestamps |

In this way, `field` should be inserted before `belongs_to`, and so on.

## GraphQL Schema

The GraphQL schema parameters need to be organized in a order of importance:

| Importance | Field |
|---|---|
| 1º | normal fields |
| 2º | custom fields (with the `resolve`) |
| 3º | associations |

In this way, `normal fields` should be inserted before `custom fields`, and so on.

Remember, the first schema rule should be respected here as well (see [here](#schemas)). Therefore, `normal fields` should be grouped and separated with a blank space. Same is valid with `custom fields` and `associations`.

[Back to top ⬆️](#table-of-contents)

> # Regular Expressions

Reviewer: I'm not sure this is the best name, because when I read "regular expressions" I immediately think of regex, which is not the case.

## Aliases

#### Ordering

The aliases should be ordered alphabetically.

```elixir
# ❌ Bad
alias Jungsoft.{
  CModule,
  BModule,
  AModule,
}

# ✅ Good
alias Jungsoft.{
  AModule,
  BModule,
  CModule,
}
```

#### SubModules

The aliases should contain only one 'sub-module'.

```elixir
# ❌ Bad
alias Jungsoft.{
  Module1.SubModule1,
  Module1.SubModule2,
  Module2.SubModule1.SubSubModule1,
  Module2.SubModule1.SubSubModule2,
}

# ✅ Good
alias Jungsoft.Module1.{
  SubModule1,
  SubModule2,
}
alias Jungsoft.Module2.SubModule1.{
  SubSubModule1,
  SubSubModule2,
}
```

[Back to top ⬆️](#table-of-contents)

> ## 'With do' Organization

Reviewer: what do you think about calling this `'With do'  indentation`?
